feat: added hinen_power component

Hinen-WeiFeng opened the pull request:
Access user-related equipment according to the interface and protocol provided by Hinen
## Breaking change
added a new component/integration for the hinen smart home 
## Proposed change

## Type of change
- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [× ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/40745
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist

- [×] The code change is tested and works locally.
- [×] Local tests pass. **Your PR cannot be merged unless tests pass**
- [×] There is no commented out code in this PR.
- [×] I have followed the [development checklist][dev-checklist]
- [×] I have followed the [perfect PR recommendations][perfect-pr]
- [×] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [×] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [×] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [×] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [×] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [×] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

To help with the load of incoming pull requests:

- [×] I have reviewed two other [open pull requests][prs] in this repository.

Hinen-WeiFeng committed:
feat: added hinen_power component

Hinen-WeiFeng committed:
fix: remove other platform and save sensor platform

Hinen-WeiFeng committed:
Merge branch 'dev' into feature-hinen-power

Hinen-WeiFeng committed:
Merge branch 'dev' into feature-hinen-power

Hinen-WeiFeng committed:
Merge branch 'dev' into feature-hinen-power

Hinen-WeiFeng committed:
Merge branch 'dev' into feature-hinen-power

Hinen-WeiFeng committed:
Merge branch 'dev' into feature-hinen-power

Hinen-WeiFeng committed:
Merge branch 'dev' into feature-hinen-power

Hinen-WeiFeng committed:
Merge branch 'dev' into feature-hinen-power

joostlek reviewed (changes requested):
All the device specific code should exist in a library hosted on PyPI.

I am not sure why there is a redirect view. Home Assistant already has an oauth redirect built in

Hinen-WeiFeng committed:
giMerge remote-tracking branch 'upstream/dev' into feature-hinen-power

Hinen-WeiFeng committed:
feat: Extract device APIs

Hinen-WeiFeng committed:
Merge branch 'feature-hinen-power' of https://github.com/Hinen-IoT/core into feature-hinen-power

Hinen-WeiFeng committed:
Merge branch 'dev' into feature-hinen-power

joostlek commented:
Yes but that doesn't explain what needs the built-in one doesn't fulfill

joostlek commented:
Well, what you are explaining is that you want users to change their redirect url via the config flow, but we already offer that feature via the my url, so I am not sure why that is a feature and why we need that redirect logic while we already have a number of OAuth integrations running existing redirect logic, which just works

joostlek commented:
Okay, but I am still missing a point, why don't you use the my step? That's not something that we dictate per integration. If a user doesn't want it, they can overwrite their default config.

Hinen-WeiFeng commented:
> Okay, but I am still missing a point, why don't you use the my step? That's not something that we dictate per integration. If a user doesn't want it, they can overwrite their default config.

I think I understand what you mean, but what we want to show users is the current implementation; I think this is related to our needs. We have also implemented the full oauth process and considered using it. Of course, if it is not allowed, please tell me

joostlek commented:
For 1: I believe we have some places where we just import a base client credential. You mention you get the client secret back, does that mean everyone has a different client secret with the same client id? for 2: I would assume that we get the full query back when we redirect from a my link.

Hinen-WeiFeng commented:
For 1, the client secret is the same, so we try not to disclose the client ID and client secret. After authorization, we call back to the home-assistant instance.
For  2, I'll try again.

Hinen-WeiFeng committed:
feat: update auth method and test

Hinen-WeiFeng committed:
Merge remote-tracking branch 'origin/feature/hinen-power-auth' into feature-hinen-power

Hinen-WeiFeng committed:
Merge branch 'dev' into feature-hinen-power

Hinen-WeiFeng committed:
fix: add user input clientId and ClientSecret

Hinen-WeiFeng committed:
Merge branch 'feature/hinen-power-client' into feature-hinen-power

Hinen-WeiFeng committed:
Merge branch 'dev' into feature-hinen-power

Hinen-WeiFeng commented:
My latest change is that users need to enter clientId and clientSecret

Hinen-WeiFeng committed:
Merge branch 'dev' into feature-hinen-power

Hinen-WeiFeng committed:
Merge branch 'dev' into feature-hinen-power

Hinen-WeiFeng committed:
Merge branch 'dev' into feature-hinen-power

Hinen-WeiFeng committed:
Merge branch 'dev' into feature-hinen-power

emontnemery commented:
@Hinen-WeiFeng I can't verify that the pypi library used by the integration meets the [transparency requirements](https://developers.home-assistant.io/docs/core/integration-quality-scale/rules/dependency-transparency?_highlight=transpare)

The [repo pointed to](https://github.com/Hinen-IoT/hinen_open_api) from the [pypi entry](https://pypi.org/project/hinen-open-api/) is a 404. Can you please take a look?

Hinen-WeiFeng commented:
@emontnemery  I have developed it as a public project, it was a private project before, you can try again

Hinen-WeiFeng committed:
Merge branch 'dev' into feature-hinen-power

Hinen-WeiFeng committed:
Merge branch 'dev' into feature-hinen-power

Hinen-WeiFeng commented:
> @Hinen-WeiFeng It looks reasonable. Can you tag a new version 1.0.1 or something to make sure the CI release flow is working? There are no runs here: https://github.com/Hinen-IoT/hinen_open_api/actions

Okay, I've dealt with it, you can take a look again

Hinen-WeiFeng committed:
fix：linter

Hinen-WeiFeng committed:
Merge branch 'dev' into feature-hinen-power

Hinen-WeiFeng committed:
feat: update hinen-open-api version 1.0.1

emontnemery commented on homeassistant/components/hinen_power/__init__.py:
```diff
@@ -0,0 +1,96 @@
+"""Support for hello auth."""
```
Shouldn't this say "Support for Hinen power"?

emontnemery commented on homeassistant/components/hinen_power/__init__.py:
```diff
@@ -0,0 +1,96 @@
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = {
+        COORDINATOR: coordinator,
+        AUTH: auth,
+    }
```
Don't store the coordinator in `hass.data`, unless the coordinator is shared between config entries (that doesn't seem to be the case?). Instead, store it in `entry.runtime_data`

emontnemery commented on homeassistant/components/hinen_power/__init__.py:
```diff
@@ -0,0 +1,96 @@
+    """Unload a config entry."""
+    if unload_ok := await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
+        hass.data[DOMAIN].pop(entry.entry_id)
+    return unload_ok
```
```suggestion
    return await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
```

emontnemery commented on homeassistant/components/hinen_power/strings.json:
```diff
@@ -0,0 +1,93 @@
+          "client_id": "OAuth Client ID",
+          "client_secret": "OAuth Client Secret",
+          "page_language": "Certification page language",
+          "region_code": "Region"
```
Why does the integration implement oauth2, but then we ask the user for these details? Isn't the point of oauth2 that the user doesn't need to fill this in?

emontnemery reviewed (changes requested):
Some initial comments

Hinen-WeiFeng committed:
fix: Adjust relevant regulations

Hinen-WeiFeng committed:
Merge branch 'dev' into feature-hinen-power

Hinen-WeiFeng committed:
Merge branch 'dev' into feature-hinen-power

Hinen-WeiFeng committed:
fix:test_init

emontnemery commented:
Linters are failing 👎 

Please set up your [development environment according to our documentation](https://developers.home-assistant.io/docs/development_environment). If you do that, the linters will run automatically when you commit which will save both you and us time.

Hinen-WeiFeng committed:
Merge branch 'dev' into feature-hinen-power

Hinen-WeiFeng committed:
fix:requirements version

Hinen-WeiFeng committed:
Merge branch 'dev' into feature-hinen-power

Hinen-WeiFeng committed:
Merge branch 'dev' into feature-hinen-power

Hinen-WeiFeng committed:
Merge branch 'dev' into feature-hinen-power

joostlek commented on homeassistant/components/hinen_power/auth_config.py:
```diff
@@ -0,0 +1,159 @@
+        ) is None:
+            secret = self.hass.data[config_entry_oauth2_flow.DATA_JWT_SECRET] = (
+                secrets.token_hex()
+            )
```
Why do we store a jwt secret in the hass.data and why in the root?

joostlek commented on homeassistant/components/hinen_power/config_flow.py:
```diff
@@ -0,0 +1,274 @@
+                {
+                    vol.Required(ATTR_AUTH_LANGUAGE, default=default_language): vol.In(
+                        dict(SUPPORTED_LANGUAGES)
+                    ),
```
Why isn't this a language selector?

joostlek commented on homeassistant/components/hinen_power/config_flow.py:
```diff
@@ -0,0 +1,274 @@
+        """Confirm reauth dialog."""
+        if user_input is None:
+            return self.async_show_form(step_id="reauth_confirm")
+        return await self.async_step_user()
```
Let's keep reauth for a followup

joostlek commented on homeassistant/components/hinen_power/config_flow.py:
```diff
@@ -0,0 +1,274 @@
+                        SelectSelectorConfig(options=selectable_devices, multiple=True)
+                    ),
+                }
+            ),
```
Why is this a step? Why don't we import all?

Hinen-WeiFeng replied:
Users may have multiple devices, and they may not need to monitor all of them; this step provides users with an option.

joostlek commented on homeassistant/components/hinen_power/config_flow.py:
```diff
@@ -0,0 +1,274 @@
+                    SelectOptionDict(value=country["code"], label=country["name"])
+                    for country in data.get("data", [])
+                ]
+            return []
```
This is a service specific code, this should exist in the PyPI library.

Hinen-WeiFeng committed:
Merge branch 'dev' into feature-hinen-power

Hinen-WeiFeng committed:
Merge branch 'home-assistant:dev' into feature-hinen-power

Hinen-WeiFeng committed:
Merge branch 'dev' into feature-hinen-power

Hinen-WeiFeng committed:
Merge branch 'dev' into feature-hinen-power

Weifeng-31 committed:
Refactor(hinen_power): enhance OAuth2 flow, improve error handling, and update sensor configurations

Weifeng-31 committed:
update(hinen_power): update test

Weifeng-31 committed:
refactor(hinen_power): Update dependencies and refactor country list fetching logic

- Upgrade hinen-open-api dependency from 1.0.1 to 1.0.2
- Remove unused enum.py file
- Refactor country list fetching logic using HinenPublic class

Hinen-WeiFeng committed:
Merge branch 'home-assistant:dev' into feature-hinen-power

Hinen-WeiFeng commented on homeassistant/components/hinen_power/enum.py:
```diff
@@ -0,0 +1,41 @@
+    """Hinen device alert status."""
+
+    NORMAL = 0
+    ALARM = 1
```
Yes

Hinen-WeiFeng committed:
Merge branch 'dev' into feature-hinen-power

Hinen-WeiFeng committed:
fix(hinen_power): resolve some issue

Hinen-WeiFeng committed:
Merge branch 'dev' into feature-hinen-power

Hinen-WeiFeng commented on homeassistant/components/hinen_power/config_flow.py:
```diff
@@ -0,0 +1,261 @@
+                title=self._title,
+                data=self._data,
+                options=user_input,
+            )
```
enforce it in the schema :  Choose at least one

Hinen-WeiFeng committed:
fix(hinen_power): resolve some issue

Hinen-WeiFeng committed:
Merge branch 'dev' into feature-hinen-power

Hinen-WeiFeng committed:
Merge branch 'dev' into feature-hinen-power

Hinen-WeiFeng committed:
Merge branch 'dev' into feature-hinen-power

Hinen-WeiFeng committed:
Remove unnecessary 'from __future__ import annotations' (Python 3.14+)

Hinen-WeiFeng committed:
Merge branch 'home-assistant:dev' into feature-hinen-power

Hinen-WeiFeng committed:
Merge branch 'dev' into feature-hinen-power

Hinen-WeiFeng committed:
Merge branch 'dev' into feature-hinen-power

Hinen-WeiFeng committed:
Enhance Hinen Power integration with improved error handling and testing

- Added OAuth2 error handling in auth_config.py for transient and reauthentication errors.
- Updated sensor.py to ensure device availability checks are robust.
- Enhanced test coverage in test_config_flow.py for API error scenarios.
- Introduced parameterized tests in test_init.py for mapping HTTP errors to OAuth2 exceptions.

Hinen-WeiFeng committed:
Merge branch 'dev' into feature-hinen-power

Hinen-WeiFeng commented on homeassistant/components/hinen_power/config_flow.py:
```diff
@@ -0,0 +1,266 @@
+from .coordinator import HinenPowerConfigEntry
+
+
+class OAuth2FlowHandler(
+    config_entry_oauth2_flow.AbstractOAuth2FlowHandler, domain=DOMAIN
+):
```
Follow-up PR processing

Hinen-WeiFeng committed:
Merge branch 'home-assistant:dev' into feature-hinen-power

Hinen-WeiFeng committed:
Merge branch 'dev' into feature-hinen-power

Hinen-WeiFeng committed:
refactor(hinen_power): Enhance Hinen Power integration with reauthentication flow, improved error handling, and updated documentation

Hinen-WeiFeng committed:
Merge branch 'dev' into feature-hinen-power

Hinen-WeiFeng committed:
Merge remote-tracking branch 'upstream/dev' into feature-hinen-power

Hinen-WeiFeng committed:
Merge branch 'dev' into feature-hinen-power

Hinen-WeiFeng committed:
Merge branch 'dev' into feature-hinen-power

Hinen-WeiFeng committed:
Merge branch 'dev' into feature-hinen-power

Hinen-WeiFeng committed:
Merge branch 'dev' into feature-hinen-power
